### ISSUE TITLE
[SR-4014] Can't Call Overloaded Method Within Another Overloaded Method

### ISSUE THREAD

**Issue by swift-ci** [swiftlang/swift#46599]
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-4014      |
   |Radar             | None         |
   |Original Reporter | Andy (JIRA User)      |
   |Type              | Bug    |

Attachment: [Download](https://user-images.githubusercontent.com/2727770/164961915-ecb2ee79-9249-47c4-9ee2-a7a942b1ff17.gz)

      
<details>
  <summary>Environment</summary>

Xcode Version 8.3 beta 3 (8W132p)  
Xcode Version 8.2.1 (8C1002)  
macOS Sierra 10.12.3

</details>

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Compiler    |
   |Labels            | Bug, DiagnosticsQoI, TypeChecker        |
   |Assignee          | None      |
   |Priority          | Medium      |

   

   md5: 076e7ba1d75c403b3cc508d17be38368

  </details>

**relates to**:  
* [SR-2657](https://bugs.swift.org/browse/SR-2657) error: '(f: () -> ())' is not convertible to '(f: () -> ())'  

**Issue Description:**

It looks like Swift is currently having issues calling overloaded methods within overloaded methods.

Case in point:

I have the following method:

``` none
    @available(*, introduced: 1.0) final public func activity(_ name: String, page: Int = 1, completionHandler completion: @escaping Anilist.API.APIArrayResponseHandler<Activity>) {
        log("Getting activity info for \(name), page \(page)")
        let usableUsername = name.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? ""
        let path = "user/\(usableUsername)/activity"
        self.associatedSession?.api.performAPITaskGET(path: path, getParameters: ["page" : "\(page)"], completionHandler: nil, completionHandlerArray: completion)
    }
```

This method has the signature \`activity(\_ name: String, page: Int, completionHandler: Anilist.API.APIArrayResponseHandler\<Activity\>\`. We will call this the "String version of Activity".

I have this other method which is simply a convenience method to call the above one with an integer as the "name" parameter.

``` none
    @available(*, introduced: 1.0) final public func activity(_ id: Int, page: Int = 0, completionHandler completion: Anilist.API.APIArrayResponseHandler<Activity>) {
        self.activity("\(id)", page: page, completionHandler: completion)
    }
```

Which like you can see, is just an overload of the method I posted earlier. \`activity(id: Int, page: Int, completionHandler: Anilist.API.APIArrayResponseHandler\<Activity\>\`. We will call this "Int version of Activity".

On the line where I try to call the "String version of Activity", Swift complains:

"Cannot convert value of type 'String' to expected argument of type 'Int'"

Basically, this method call wants to call itself recursively and does not see there's a different version in the same scope.

Unfortunately, I have no idea when this broke. I wrote this code in the Xcode 8.0/Swift 3.0 beta days and it used to work fine. If memory serves me right, this error started happening on the final release of Xcode 8.0 and Swift 3.0

I need to note that autocompletion works fine. When typing \`self.activity...\` within the "Int Version of Activity" autocompletion shows and completes the "String version of Activity" and inserts it there. So this is most likely a compiler bug.

**Comment by slavapestov** [comment]
Looks like this has since been fixed:

```
bad.swift:10:63: error: passing non-escaping parameter 'completion' to function expecting an '@escaping' closure
 7 |     }
 8 | 
 9 |     final public func activity(_ id: Int, page: Int = 0, completionHandler completion: G<Activity>) {
   |                                                                            `- note: parameter 'completion' is implicitly non-escaping
10 |         self.activity("\(id)", page: page, completionHandler: completion)
   |                                                               `- error: passing non-escaping parameter 'completion' to function expecting an '@escaping' closure
11 |     }
12 | }
```
